Sources/HTTP/TCP.swift

```diff
@@ -0,0 +1,713 @@
+    var isWriting = false
+
+    //// Close handler
+    typealias OnClose = () -> ()
+
+    /// Called when this client closes
+    var onClose: OnClose?
+
+    /// Creates a new Remote Client from the a socket
+    init(socket: TCPSocket, on eventLoop: DispatchQueue) {
+        self.socket = socket
+        self.eventLoop = eventLoop
+
+        // Allocate one TCP packet
+        let size = 1024 // 65_507
+        let pointer = UnsafeMutablePointer<UInt8>.allocate(capacity: size)
+        self.outputBuffer = UnsafeMutableBufferPointer<UInt8>(start: pointer, count: size)
+    }
+
+    convenience init(on eventLoop: DispatchQueue) throws {
+        let socket = try TCPSocket()
+        socket.disablePipeSignal()
+        self.init(socket: socket, on: eventLoop)
+    }
+
+    /// See InputStream.onInput
+    func write(_ input: UnsafeBufferPointer<UInt8>) {
+        do {
+            let count = try socket.write(from: input)
+
+            guard count == input.count else {
+                let data = Data(input[input.count...])
+
+                inputBuffer.append(data)
+                ensureWriteSourceResumed()
+                return
+            }
+        } catch {
+            inputBuffer.append(Data(input))
+            ensureWriteSourceResumed()
+        }
+    }
+
+    /// Handles DispatchData input
+    func write(_ input: DispatchData) {
+        inputBuffer.append(Data(input))
+        ensureWriteSourceResumed()
+    }
+
+    /// Handles Data input
+    func write(_ input: Data) {
+        inputBuffer.append(input)
+        ensureWriteSourceResumed()
+    }
+
+    private func ensureWriteSourceResumed() {
+        if !isWriting {
+            ensureWriteSource().resume()
+            isWriting = true
+        }
+    }
+
+    /// Creates a new WriteSource is there is no write source yet
+    private func ensureWriteSource() -> DispatchSourceWrite {
+        guard let source = writeSource else {
+            let source = DispatchSource.makeWriteSource(
+                fileDescriptor: socket.descriptor,
+                queue: eventLoop
+            )
+
+            source.setEventHandler {
+                // grab input buffer
+                guard self.inputBuffer.count > 0 else {
+                    self.writeSource?.suspend()
+                    return
+                }
+
+                let data = self.inputBuffer.removeFirst()
+
+                if self.inputBuffer.count == 0 {
+                    // important: make sure to suspend or else writeable
+                    // will keep calling.
+                    self.writeSource?.suspend()
+
+                    self.isWriting = false
+                }
+
+                data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
+                    let buffer = UnsafeBufferPointer<UInt8>(start: pointer, count: data.count)
+
+                    do {
+                        let length = try self.socket.write(from: buffer)
```

Comment:
I think you have a low-hanging fruit here. Instead of doing individual writes, use `writev` w/ the array of buffers.